src/vscode-bicep/src/language/protocol.ts

```diff
@@ -55,3 +57,14 @@ export const bicepCacheRequestType = new ProtocolRequestType<
   void,
   void
 >("textDocument/bicepCache");
+
+export interface InsertResourceParams {
+  textDocument: TextDocumentIdentifier;
+  position: Position;
+  resourceId: string;
+}
+
+export const insertResourceRequestType = new ProtocolNotificationType<
```

Comment:
>ProtocolNotificationType [](http://example.com/codeflow?start=45&length=24)

Shouldn't this be a request rather than a notification, so failures can be reported back to the client?

Reply:
Known error cases are currently reported to the client via `showError`. Unhandled exceptions currently ignored - I could either add an outer handling block for them, or implement request.

What I don't like about using request and allowing errors to be thrown is the error is pretty intrusive and unhelpful, so it probably still makes sense to handle 'known' failure cases with `showError`:
![image](https://user-images.githubusercontent.com/38542602/141118135-d18522de-4abd-441f-8a7b-d8dedf5033e4.png)

Reply:
Can leave as is. The error experience looks really good and if we discover a reason, we can always switch.

Reply:
BTW, the O# server allows you to register a generic unhandled exception handler (`server.WithUnhandledExceptionHandler(...)`). Might be worth us adding some telemetry there and/or formatting unhandled exceptions in a more user-friendly manner with a "report issue" button. I'll raise an issue for this.